File: codeChallenges/hackerRank/mathematics/leoPrime.go
package main

import (
    "fmt"
)

func primeGen() (func() int) {
  prime := 2
  return func() int {
    last := prime
    for num := last + 1; ; num++{
      isPrime := true
      for i := 2; i < num; i++ {
        if num % i == 0 {
          isPrime = false
        }
      }
      if isPrime {
        prime = num
        break
      }
    }
    return last
  }
}

func maxFactors(n int) int {
  if n <= 1 {
    return 0
  }
  max, sum := 0, 1
  primes := primeGen()
  for sum <= n {
    sum *= primes()
    max++
  }
  return max - 1
}

func main() {
  var lines, n int
  fmt.Scan(&lines)
  for lines > 0 {
    fmt.Scan(&n)
    fmt.Println(maxFactors(n))
    lines--
  }
}

File: codeChallenges/hackerRank/mathematics/leoPrime.js
/*
given n
find max number of unique primes of any number
in the range [1 ... n] inclusive both sides
*/

const Primes = function*(start = 2) {
  for (let i = start; true; i++) {
    let isPrime = true;
    for (let j = 2; j < i; j++) {
      if (i % j === 0) {
        isPrime = false;
        break;
      }
    }
    if (isPrime) {
      yield i;
    }
  }
};

const m = n => {
  const p = Primes();
  if (n <= 1) {
    return 0;
  }
  let maxFactors = 0;
  let sum = 1;
  while (sum <= n) {
    let nextPrime = p.next().value;
    sum *= nextPrime;
    maxFactors++;
  }
  return maxFactors - 1;
};

function processData(input) {
  const str = input.split("\n");
  const lines = Number(str[0]);
  for (let cases = 0; cases < lines; cases++) {
    let n = Number(str[cases + 1]);
    console.log(m(n));
  }
}
// console.log(m(500));
